src/esbmc/bmc.cpp

```diff
@@ -658,3 +700,163 @@ smt_convt::resultt bmct::run_thread(std::shared_ptr<symex_target_equationt> &eq)
+
+      const bool fail_fast = options.get_bool_option("multi-fail-fast");
+      // This loop is mainly for fail-fast.
+      try
+      {
+        while(!solver_job.joinable())
+        {
+          // Try again 100ms later
+          using namespace std::chrono_literals;
+          std::this_thread::sleep_for(100ms);
+          // Did someone finished already?
+          if(fail_fast && final_result == smt_convt::P_SATISFIABLE)
+          {
+            log_status("Other thread already found a SAT VCC.");
+            throw 0;
+          }
+        }
+        solver_job.join();
+        report_multi_property_trace(result, claim.claim_msg);
+        if(result == smt_convt::P_SATISFIABLE)
+        {
+          const std::lock_guard<std::mutex> lock(result_mutex);
+          // First, check if someone else find the solution!
+          if(fail_fast && final_result == smt_convt::P_SATISFIABLE)
+          {
+            log_status(
+              "Found solution for VCC. But, other thread found it first.");
+            throw 0;
+          }
+          goto_tracet goto_trace;
+          build_goto_trace(local_eq, runtime_solver, goto_trace, false);
+          // TODO: Replace this with a test-case for coverage!
+          std::string output_file = options.get_option("cex-output");
+          if(output_file != "")
+          {
+            std::ofstream out(fmt::format("{}-{}", output_file, ce_counter++));
+            show_goto_trace(out, ns, goto_trace);
+          }
+          std::ostringstream oss;
+          oss << "\n" << ansi_green << "[Counterexample]" << ansi_clr << "\n";
+          show_goto_trace(oss, ns, goto_trace);
+          log_result("{}", oss.str());
+          final_result = result;
+        }
+        // TODO: This is the place to store into a cache
+      }
+      catch(...)
+      {
+        log_status("Failing Fast");
+      }
+    };
+
+  // PARALLEL
+  if(options.get_bool_option("parallel-solving"))
+  {
+    /* NOTE: I would love to use std::for_each here, but it is not giving
+       * the result I would expect. My guess is either compiler version
+       * or some magic flag that we are not using.
+       *
+       * Nevertheless, we can achieve the same results by just creating
+       * threads.
+       */
+
+    // TODO: Running everything in parallel might be a bad idea.
+    //       Should we also add a thread pool?
+    std::vector<std::thread> parallel_jobs;
+    for(const auto &i : jobs)
+      parallel_jobs.push_back(std::thread(job_function, i));
+
+    // Main driver
+    size_t not_finished = parallel_jobs.size();
+    while(not_finished)
+    {
+      for(auto &t : parallel_jobs)
+      {
+        if(t.joinable())
+        {
+          t.join();
+          // There is no data-race in this loop, we should be fine.
+          not_finished--;
+        }
+      }
+      // We could filter out parallel_jobs for the ones that were already joined.
```

Comment:
Hm, guess I did a bad English there:

"We could remove joined jobs from the parallel_jobs vector"